replica/table.cc

```diff
@@ -1254,12 +1255,13 @@ void table::set_compaction_strategy(sstables::compaction_strategy_type strategy)
 
             new_sstables = make_lw_shared<sstables::sstable_set>(new_cs.make_sstable_set(t._schema));
             cg.main_sstables()->for_each_sstable([this] (const sstables::shared_sstable& s) {
-                add_sstable_to_backlog_tracker(new_bt, s);
                 new_sstables->insert(s);
+                new_sstables_for_backlog_tracker.push_back(s);
             });
         }
 
         void execute() noexcept {
+            new_bt.replace_sstables({}, std::move(new_sstables_for_backlog_tracker));
```

Comment:
new_sstables_for_backlog_tracker can also be computed on-the-fly here using `boost::copy_range<sstables::shared_sstable>(*new_sstables->all)`, no?
It might be slightly more efficient since it will reserve the required capacity once.

Reply:
This is noexcept path though. also not a hot path. I can reserve new_sstables_for_backlog_tracker upfront. what do you think?

Reply:
> This is noexcept path though.

ok, so let's do that in `prepare`.

> also not a hot path. I can reserve new_sstables_for_backlog_tracker upfront. what do you think?

Do we know the size upfront?
I don't think sstable_set gives you that, does it?

Reply:
> > This is noexcept path though.
> 
> ok, so let's do that in `prepare`.
> 
> > also not a hot path. I can reserve new_sstables_for_backlog_tracker upfront. what do you think?
> 
> Do we know the size upfront? I don't think sstable_set gives you that, does it?

I think we can easily implement a size() method for sstable_set. Today, we have to call all() which can copy elements, but that's possibly better than reallocating log base_2 N, where N is number of elements inserted into container.

Reply:
compaction_backlog_tracker::replace_sstables() can fail, no?

Reply:
we disable tracker on failure, of course, that's an implementation detail. but if we want interface to reflect it, we need to make it noexcept. 

thinking better, we can just call replace_sstables in the end of prepare(), and only wire new tracker in execute(). thanks.

Reply:
> > > This is noexcept path though.
> > 
> > 
> > ok, so let's do that in `prepare`.
> > > also not a hot path. I can reserve new_sstables_for_backlog_tracker upfront. what do you think?
> > 
> > 
> > Do we know the size upfront? I don't think sstable_set gives you that, does it?
> 
> I think we can easily implement a size() method for sstable_set. Today, we have to call all() which can copy elements, but that's possibly better than reallocating log base_2 N, where N is number of elements inserted into container.

Let's consider this for follow up then.

Reply:
> compaction_backlog_tracker::replace_sstables() can fail, no?

Yup, https://github.com/scylladb/scylladb/blob/54170749b853e166223f865621732be4b527ffe0/compaction/compaction_manager.cc#L1725

Why does `execute` need to be noexcept?

Reply:
#10297

Reply:
> > > > This is noexcept path though.
> > > 
> > > 
> > > ok, so let's do that in `prepare`.
> > > > also not a hot path. I can reserve new_sstables_for_backlog_tracker upfront. what do you think?
> > > 
> > > 
> > > Do we know the size upfront? I don't think sstable_set gives you that, does it?
> > 
> > 
> > I think we can easily implement a size() method for sstable_set. Today, we have to call all() which can copy elements, but that's possibly better than reallocating log base_2 N, where N is number of elements inserted into container.
> 
> Let's consider this for follow up then.

sure.